File: reminders.py
"""Слой 3: контент напоминания.

Собирает конкретный текст напоминания и три кнопки действия. Каждое
напоминание = задание + прямая ссылка + оценка времени + [Готово][Позже][Пропустить].

CSCA присылается как ПОДРОБНЫЙ урок: сначала на английском, потом на русском,
потом практика. Текст урока математический (со знаками ^, _, *, /), поэтому
идёт БЕЗ Markdown, чтобы Telegram не поломал разметку. Длинный урок бьётся на
несколько сообщений — клавиатура вешается на последнее.
"""

from telegram import InlineKeyboardButton, InlineKeyboardMarkup

import config
import content

# Практический лимит длины сообщения Telegram — 4096. Держим запас.
CHUNK_LIMIT = 3500


def split_chunks(text: str, limit: int = CHUNK_LIMIT):
    """Бьёт длинный текст на части по границам абзацев (не рвёт абзац)."""
    paras = text.split("\n\n")
    chunks, cur = [], ""
    for p in paras:
        piece = (("\n\n" + p) if cur else p)
        if len(cur) + len(piece) <= limit:
            cur += piece
        else:
            if cur:
                chunks.append(cur)
            # Абзац сам по себе длиннее лимита — режем жёстко.
            while len(p) > limit:
                chunks.append(p[:limit])
                p = p[limit:]
            cur = p
    if cur:
        chunks.append(cur)
    return chunks or [text]


def build_reminder(track_id: str, content_index: int):
    """Возвращает dict: chunks (list[str]), link, title, minutes, markdown (bool).

    markdown=False означает «слать без parse_mode» (для математических уроков).
    """
    if track_id == config.TRACK_6MIN:
        number, title, link, minutes = content.six_min_task(content_index)
        text = (
            f"🎧 *6 Minute English #{number}*\n"
            f"«{title}»\n\n"
            f"Послушать + разобрать словарь. ~{minutes} мин."
        )
        return {"chunks": [text], "link": link, "title": title,
                "minutes": minutes, "markdown": True}

    if track_id == config.TRACK_DET:
        number, title, desc, tip, link, minutes = content.det_task(content_index)
        text = (
            f"🦉 *Duolingo Test #{number}*\n"
            f"*{title}*\n\n"
            f"{desc}\n\n"
            f"💡 {tip}\n\n"
            f"~{minutes} мин. Практика — по кнопке ниже."
        )
        return {"chunks": [text], "link": link, "title": title,
                "minutes": minutes, "markdown": True}

    if track_id == config.TRACK_CSCA:
        block_no, title, body, link, minutes = content.csca_task(content_index)
        total = len(content.CSCA_LESSONS)
        full = (
            f"📐 CSCA Math · тема {block_no}/{total} (~{minutes} мин)\n"
            f"{title}\n\n"
            f"{body}"
        )
        return {"chunks": split_chunks(full), "link": link, "title": title,
                "minutes": minutes, "markdown": False}

    if track_id == config.TRACK_EGE:
        title, link, minutes = content.ege_task(content_index)
        text = (
            f"🧮 *Пробник ЕГЭ (профиль)*\n"
            f"{title}\n\n"
            f"Полный вариант по таймеру, ~{minutes} мин. "
            f"После — разбор ошибок."
        )
        return {"chunks": [text], "link": link, "title": title,
                "minutes": minutes, "markdown": True}

    return {"chunks": ["Напоминание"], "link": "", "title": "task",
            "minutes": 0, "markdown": True}


def action_keyboard(hid: int, link: str, link_label="Открыть"):
    """Кнопки: ссылка + Готово / Позже / Пропустить.

    'Позже' раскрывается в под-меню (+2ч / завтра) через отдельный callback.
    """
    return InlineKeyboardMarkup([
        [InlineKeyboardButton(f"🔗 {link_label}", url=link)],
        [
            InlineKeyboardButton("✅ Готово", callback_data=f"done:{hid}"),
            InlineKeyboardButton("🕑 Позже", callback_data=f"later:{hid}"),
            InlineKeyboardButton("⏭ Пропустить", callback_data=f"skip:{hid}"),
        ],
    ])


def later_keyboard(hid: int):
    return InlineKeyboardMarkup([
        [
            InlineKeyboardButton("+2 часа", callback_data=f"post2h:{hid}"),
            InlineKeyboardButton("Завтра", callback_data=f"posttom:{hid}"),
        ],
        [InlineKeyboardButton("« назад", callback_data=f"back:{hid}")],
    ])


def build_skip_question(track_id: str, hid: int):
    """Правило пропусков: после 2 пропусков подряд — не напоминание, а вопрос."""
    title = config.TRACK_TITLES[track_id]
    text = (
        f"🤔 *{title}* не идёт уже 2 раза подряд.\n"
        f"Что случилось?"
    )
    kb = InlineKeyboardMarkup([
        [InlineKeyboardButton("😮‍💨 Тяжело", callback_data=f"why:{hid}:hard")],
        [InlineKeyboardButton("⏳ Нет времени", callback_data=f"why:{hid}:notime")],
        [InlineKeyboardButton("😐 Надоел формат", callback_data=f"why:{hid}:bored")],
    ])
    return text, kb


def why_response(track_id: str, reason: str):
    """Ответ бота на причину пропусков + предложение по адаптации."""
    title = config.TRACK_TITLES[track_id]
    if reason == "hard":
        return (
            f"Понял. Сделаем *{title}* легче: бери половину задания и просто "
            f"отметь «Готово». Полдела — уже дело. В следующий раз пришлю как обычно."
        )
    if reason == "notime":
        return (
            f"Ок, времени нет. Могу сдвинуть *{title}* на другое время или сделать "
            f"реже. Напиши /tracks — там поменяем расписание за 10 секунд."
        )
    if reason == "bored":
        return (
            f"Формат приелся — нормально. По *{title}* переключу на следующий блок "
            f"программы, будет другое. Дам передышку на пару дней."
        )
    return "Принял."

File: worker/src/reminders.js
// Слой 3: сборка текста напоминания и клавиатур.
// CSCA — подробный урок (EN -> RU -> практика), шлётся БЕЗ Markdown (математика
// со знаками ^, _, *), при необходимости бьётся на несколько сообщений.

import { TRACK_6MIN, TRACK_EGE, TRACK_CSCA, TRACK_DET, TRACK_TITLES } from "./config.js";
import { sixMinTask, cscaTask, egeTask, detTask } from "./content.js";

const CHUNK_LIMIT = 3500;

export function splitChunks(text, limit = CHUNK_LIMIT) {
  const paras = text.split("\n\n");
  const chunks = [];
  let cur = "";
  for (let p of paras) {
    const piece = cur ? "\n\n" + p : p;
    if (cur.length + piece.length <= limit) {
      cur += piece;
    } else {
      if (cur) chunks.push(cur);
      while (p.length > limit) {
        chunks.push(p.slice(0, limit));
        p = p.slice(limit);
      }
      cur = p;
    }
  }
  if (cur) chunks.push(cur);
  return chunks.length ? chunks : [text];
}

// Возвращает { chunks, link, title, minutes, markdown }.
export function buildReminder(trackId, contentIndex) {
  if (trackId === TRACK_6MIN) {
    const t = sixMinTask(contentIndex);
    const text = `🎧 *6 Minute English #${t.number}*\n«${t.title}»\n\nПослушать + разобрать словарь. ~${t.minutes} мин.`;
    return { chunks: [text], link: t.link, title: t.title, minutes: t.minutes, markdown: true };
  }
  if (trackId === TRACK_DET) {
    const t = detTask(contentIndex);
    const text = `🦉 *Duolingo Test #${t.number}*\n*${t.title}*\n\n${t.desc}\n\n💡 ${t.tip}\n\n~${t.minutes} мин. Практика — по кнопке ниже.`;
    return { chunks: [text], link: t.link, title: t.title, minutes: t.minutes, markdown: true };
  }
  if (trackId === TRACK_CSCA) {
    const t = cscaTask(contentIndex);
    const full =
      `📐 CSCA Math · блок ${t.block}\n${t.title}\n\n` +
      `🇬🇧 ENGLISH\n${t.english}\n\n` +
      `🇷🇺 РУССКИЙ\n${t.russian}\n\n` +
      `▶ Практика (~${t.minutes} мин)\n${t.task}`;
    return { chunks: splitChunks(full), link: t.link, title: t.title, minutes: t.minutes, markdown: false };
  }
  if (trackId === TRACK_EGE) {
    const t = egeTask(contentIndex);
    const text = `🧮 *Пробник ЕГЭ (профиль)*\n${t.title}\n\nПолный вариант по таймеру, ~${t.minutes} мин. После — разбор ошибок.`;
    return { chunks: [text], link: t.link, title: t.title, minutes: t.minutes, markdown: true };
  }
  return { chunks: ["Напоминание"], link: "", title: "task", minutes: 0, markdown: true };
}

export function actionKeyboard(hid, link, label = "Открыть") {
  return {
    inline_keyboard: [
      [{ text: `🔗 ${label}`, url: link }],
      [
        { text: "✅ Готово", callback_data: `done:${hid}` },
        { text: "🕑 Позже", callback_data: `later:${hid}` },
        { text: "⏭ Пропустить", callback_data: `skip:${hid}` },
      ],
    ],
  };
}

export function laterKeyboard(hid) {
  return {
    inline_keyboard: [
      [
        { text: "+2 часа", callback_data: `post2h:${hid}` },
        { text: "Завтра", callback_data: `posttom:${hid}` },
      ],
      [{ text: "« назад", callback_data: `back:${hid}` }],
    ],
  };
}

export function buildSkipQuestion(trackId, hid) {
  const title = TRACK_TITLES[trackId];
  const text = `🤔 *${title}* не идёт уже 2 раза подряд.\nЧто случилось?`;
  const kb = {
    inline_keyboard: [
      [{ text: "😮‍💨 Тяжело", callback_data: `why:${hid}:hard` }],
      [{ text: "⏳ Нет времени", callback_data: `why:${hid}:notime` }],
      [{ text: "😐 Надоел формат", callback_data: `why:${hid}:bored` }],
    ],
  };
  return { text, kb };
}

export function whyResponse(trackId, reason) {
  const title = TRACK_TITLES[trackId];
  if (reason === "hard")
    return `Понял. Сделаем *${title}* легче: бери половину задания и просто отметь «Готово». Полдела — уже дело. В следующий раз пришлю как обычно.`;
  if (reason === "notime")
    return `Ок, времени нет. Могу сдвинуть *${title}* на другое время или сделать реже. Напиши /tracks — там поменяем расписание за 10 секунд.`;
  if (reason === "bored")
    return `Формат приелся — нормально. По *${title}* переключу на следующий блок программы, будет другое. Дам передышку на пару дней.`;
  return "Принял.";
}
